feat: Job Emblems

pacampbell opened the pull request:
- Added support for job emblems using the menus at Renton.
- Created new handlers and packets for the new menus.
- Created new personal quests to unlock emblems.

# Checklist:
- [x] The project compiles
- [x] The PR targets `develop` branch

pacampbell committed:
feat: Job Emblems

- Added support for job emblems
- Created new personal quests to unlock emblems.
- Added support for emblems using the menus at Renton.
- Created new handlers and packets for the new menus.

RyanYappert commented on Arrowgene.Ddon.Database/Sql/Core/DdonSqlDbCharacterCommon.cs:
```diff
@@ -157,6 +157,7 @@ private void QueryCharacterCommonData(DbConnection conn, CharacterCommon common)
 
                                 using DbConnection connection3 = OpenNewConnection();
                                 item.AddStatusParamList = GetEquipmentLimitBreakRecord(item.UId, connection3);
+                                item.EquipStatParamList.Add(new CDataEquipStatParam() { EffectID = (byte) EquipStatId.MaxHp, EffectValue = 50 });
```
👀

pacampbell replied:
Oops. It doesn't do anything though but should fix it :)

RyanYappert commented on Arrowgene.Ddon.GameServer/Characters/CharacterManager.cs:
```diff
@@ -471,7 +473,7 @@ private PacketQueue NotifyClientOfCharacterStatus(GameClient client, CharacterCo
             {
                 S2CContextGetLobbyPlayerContextNtc ntc1 = new S2CContextGetLobbyPlayerContextNtc();
-                GameStructure.S2CContextGetLobbyPlayerContextNtc(ntc1, (Character) characterCommon);
+                GameStructure.S2CContextGetLobbyPlayerContextNtc(Server, ntc1, (Character) characterCommon);
```
I thought we were gonna pull these out and make them methods of the `Character` object? `public S2CContextGetLobbyPlayerContextNtc ToLobbyPlayerContextNtc(DdonGameServer server)` or something.

pacampbell replied:
This thing is a lot harder to do because it's used in enough places. We should do it in a followup PR.

RyanYappert commented on Arrowgene.Ddon.GameServer/Characters/JobEmblemManager.cs:
```diff
@@ -0,0 +1,183 @@
+            emblemItem.EquipStatParamList = GetEquipStatParamList(character.JobEmblems[jobId]);
+        }
+
+        private byte GetInheritenceBasePercentageChance(Item item)
```
Typo, Inherit**e**nce > Inherit**a**nce, in multiple places.

RyanYappert commented on Arrowgene.Ddon.GameServer/Characters/PlayPointManager.cs:
```diff
@@ -60,45 +61,44 @@ public S2CJobUpdatePlayPointNtc AddPlayPoint(GameClient client, (uint BasePoints
+            client.Send(RemovePlayPoint2(client, jobId.Value, removedPoints, type));
+        }
+
+        public S2CJobUpdatePlayPointNtc RemovePlayPoint2(GameClient client, JobId jobId, uint amount, byte type = 0, DbConnection? connectionIn = null)
```
Is there a reason we have two `RemovePlayPoint`s? We should probably just have the single instance that returns the packet and send it from the call site (seemingly just `JobJobValueShopBuyItemHandler` prior to this), rather than being unclear about which one sends and which doesn't.

While we're at it, we can move `JobJobValueShopBuyItemHandler` into a transaction.

pacampbell replied:
I'm trying to not change so many things.

RyanYappert commented on Arrowgene.Ddon.Server/Settings/EmblemSettings.cs:
```diff
@@ -0,0 +1,493 @@
+        /// Defines the maximum amount you can increase your odds when attaching crests to the job emblem.
+        /// </summary>
+        [DefaultValue(_MaxInhertienceChanceIncrease)]
+        public byte MaxInhertienceChanceIncrease
```
See above, this time with Inhertience > Inheritance

alborrajo commented on Arrowgene.Ddon.Database/DdonDatabaseBuilder.cs:
```diff
@@ -13,7 +13,7 @@ public static class DdonDatabaseBuilder
     private const string DefaultSchemaFile = "Script/schema_sqlite.sql";
 
-    public const uint Version = 41;
+    public const uint Version = 42;
```
PR #822 uses the same version, we have to be careful here

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/Migration/00000042_JobEmblemMigration.cs:
ditto

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/Migration/DdonSqlDbJobEmblem.cs:
```diff
@@ -0,0 +1,131 @@
+        VALUES ({BuildQueryInsert(JobEmblemFields)})
+        ON CONFLICT (character_id, job_id)
+        DO UPDATE SET {BuildQueryUpdate(JobEmblemFields)};
+    """;
```
Is this query compatible with both PSQL and SQLite?

pacampbell replied:
It should be considering I copied it from a query Sam made recently.

Sehkah replied:
Note: For an optimized ON CONFLICT UPDATE you want to separate the key from non-key fields and add the "EXCLUDED." prefix for the non-key fields.

For example in the DdonSqlDbPawnCraftProgress:
```C#
    private readonly string SqlUpsertPawnCraftProgress =
        $"""
         INSERT INTO "ddon_equipment_limit_break" ({BuildQueryField(PawnCraftProgressFields)}) 
                        VALUES ({BuildQueryInsert(PawnCraftProgressFields)}) 
                        ON CONFLICT ("craft_character_id", "craft_lead_pawn_id") 
                        DO UPDATE SET {BuildQueryUpdateWithPrefix("EXCLUDED.", PawnCraftProgressNonKeyFields)};
         """;
```

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/Migration/DdonSqlDbJobEmblem.cs:
```diff
@@ -0,0 +1,131 @@
+        result.StatLevels[EquipStatId.IceResist] = GetByte(reader, "ice_resist");
+        result.StatLevels[EquipStatId.ThunderResist] = GetByte(reader, "thunder_resist");
+        result.StatLevels[EquipStatId.HolyResist] = GetByte(reader, "holy_resist");
+        result.StatLevels[EquipStatId.DarkResist] = GetByte(reader, "dark_resist");
```
Should StatLevels really be a dictionary when it gets written and read into a table with non nullable columns?

pacampbell replied:
What I have stored in the DB is not what is used in the actual packets. I need some way to take the level of the stat, then look up the value in another config table and calculate the final value.

alborrajo replied:
What i mean is, can it not be a regular field? `StatLevels.IceResist = GetByte(reader, "ice_resist")`

pacampbell replied:
I can't loop over those fields without doing reflection though.

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/JobEmblemAttachElementHandler.cs:
```diff
@@ -0,0 +1,129 @@
+            var packets = new PacketQueue();
+            var itemUpdateNtc = new S2CItemUpdateCharacterItemNtc()
+            {
+                UpdateType = ItemNoticeType.GatherEquipItem
```
Is this right? Or is this one of the cases where the update type doesn't matter

pacampbell replied:
When nothing works, this update type always fixes everything. Maybe it needs a better name but not sure what it is. It's needed for equipping items from BBM chests, otherwise it hangs. The limit break and emblem also don't show unless I use this type.

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/JobEmblemResetParamLevelHandler.cs:
```diff
@@ -0,0 +1,87 @@
+
+            var updateCharacterItemNtc = new S2CItemUpdateCharacterItemNtc()
+            {
+                UpdateType = ItemNoticeType.GatherEquipItem,
```
ditto

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/JobEmblemUpdateParamLevel.cs:
```diff
@@ -0,0 +1,75 @@
+            var packets = new PacketQueue();
+            var updateCharacterItemNtc = new S2CItemUpdateCharacterItemNtc()
+            {
+                UpdateType = ItemNoticeType.GatherEquipItem,
```
ditto

alborrajo commented on Arrowgene.Ddon.Shared/Entity/Structure/CDataTimeGainQuestUnk2.cs:
```diff
@@ -8,12 +8,12 @@ public class CDataTimeGainQuestUnk2
         {
             WalletPointList = new List<CDataWalletPoint>();
-            Unk0List = new List<CDataRequiredItem>();
+            Unk0List = new List<CDataItemAmount>();
```
The property name could be updated while at it

alborrajo commented on Arrowgene.Ddon.Shared/Entity/Structure/CDataItemEmbodyCostParam.cs:
```diff
@@ -9,28 +9,28 @@ public class CDataItemEmbodyCostParam
         {
             WalletPoints = new List<CDataWalletPoint>();
-            Unk1 = new List<CDataRequiredItem>();
+            Unk1 = new List<CDataItemAmount>();
```
ditto

pacampbell committed:
code review feedback

- Fix typo Inheritence -> Inheritance.
- Remove 50 HP added to all items in CharacterCommon query.
